### ISSUE TITLE
Reproducing Training Results

### ISSUE THREAD

**Issue by yoloboy2500** [AlexeyAB/darknet#1403]
Hi,

I'm interested in being able to closely imitate training results when I run a model twice. Right now I get fairly different results when predicting on the same images after training the exact same model twice. I have turned off all forms of data augmentation that I can within the .cfg file (including flip) and am using YoloV2 and the darknet19_448.conv.23 model to begin training. I understand that by nature this is stochastic, but if there are any possibilities to reduce randomness, help would be appreciated.

Thank you.

**Comment by yoloboy2500** [comment]
@AlexeyAB

**Comment by AlexeyAB** [comment]
@yoloboy2500 Hi,

Here the images will be loaded randomly - so you should change this function to load images sequentially: https://github.com/AlexeyAB/darknet/blob/7f9145729c2a2b64a7aeb3806f7c3d21771fce94/src/data.c#L42-L59


Try to use such function:
```cpp
 char **get_random_paths(char **paths, int n, int m) 
 { 
     char **random_paths = calloc(n, sizeof(char*)); 
     int i; 
     pthread_mutex_lock(&mutex); 
     //printf("n = %d \n", n); 
     for(i = 0; i < n; ++i){ 
         do { 
             //int index = random_gen() % m; 
             static int index = -1; 
             index = (++index) % m;
             random_paths[i] = paths[index]; 
             //if(i == 0) printf("%s\n", paths[index]); 
             //printf("grp: %s\n", paths[index]); 
             if (strlen(random_paths[i]) <= 4) printf(" Very small path to the image: %s \n", random_paths[i]); 
         } while (strlen(random_paths[i]) == 0); 
     } 
     pthread_mutex_unlock(&mutex); 
     return random_paths; 
 } 
```

**Comment by yoloboy2500** [comment]
@AlexeyAB Thank you! I will try this. Are there any areas of randomness with regards to initialization? I understand that using darknet19_448.conv.23 makes it so the first 23 weights are already initialized, but does the final layer begin with random weights?

**Comment by AlexeyAB** [comment]
@yoloboy2500 Yes, it is there: https://github.com/AlexeyAB/darknet/blob/140bc1882dd48dfff93e94dae0fb780351513c9e/src/convolutional_layer.c#L285

You should change this line too.

**Comment by yoloboy2500** [comment]
Hi again @AlexeyAB,

I implemented these changes. I changed rand_uniform(-1,1) to 0.001 and implemented your suggested change to the image loading order. I still received very different results after the same number of iterations when training twice. Do you have any other suggestions or thoughts where this might be coming from?

Thank you.

**Comment by AlexeyAB** [comment]
Hi,

Also set `args.threads = 1; // 16` https://github.com/AlexeyAB/darknet/blob/7f9145729c2a2b64a7aeb3806f7c3d21771fce94/src/detector.c#L105

And check one more time that you disabled data augmentation.

**Comment by yoloboy2500** [comment]
@AlexeyAB 

I will try this. To disable data augmentation, I simply set the various parameters = to 0 in the .cfg file. This is my .cfg:

[net]
batch=64
subdivisions=8
height=416
width=416
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 0
exposure = 0
hue=0
flip=0

learning_rate=0.0001
max_batches = 120000
policy=steps
steps=100,25000,35000
scales=10,.1,.1

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky


#######

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[route]
layers=-9

[reorg]
stride=2

[route]
layers=-1,-3

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=55
activation=linear

[region]
anchors = 1.08,1.19,  3.42,4.41,  6.63,11.38,  9.42,5.11,  16.62,10.52
bias_match=1
classes=6
coords=4
num=5
softmax=1
jitter=0
rescore=1

object_scale=5
noobject_scale=1
class_scale=1
coord_scale=1

absolute=1
thresh = .6
random=0

To ensure they are disabled, should I also disable them on the backend?

**Comment by AlexeyAB** [comment]
should be

```
saturation = 1
exposure = 1
hue = 0
flip = 0

jitter=0
random=0
```

**Comment by yoloboy2500** [comment]
Hi again @AlexeyAB 

I changed saturation to 1 and exposure to 1, and set the args.threads to equal 1. Trying to reproduce training again lead to very different results after the same number of iterations. It seems that this is happening early on in the training, possibly causing an immediate divergence, because the average loss is different on the first iteration if I start training the exact same model twice, in the same command prompt. Do you know if any other initialization randomness would be causing an immediate change during the beginning of training?

Thanks.
